db/db_impl/db_impl_write.cc

```diff
@@ -1309,9 +1303,15 @@ Status DBImpl::HandleWriteBufferFull(WriteContext* write_context) {
       if (cfd->IsDropped()) {
         continue;
       }
-      if (!cfd->mem()->IsEmpty()) {
+      if (!cfd->mem()->IsEmpty() &&
+          cfd->mem()->ApproximateMemoryUsageFast() >
+              cfd->imm()->ApproximateUnflushedMemTablesMemoryUsage()) {
```

Comment:
after experimenting with it a bit today I realized this heuristic is confusing. Instead of this, how about only let write buffer manager trigger flush when `cfd->imm()->ApproximateUnflushedMemTablesMemoryUsage() == 0`?

Reply:
Done.

Reply:
This still isn't my favorite heuristic though. IMO it'd be good if (1) we permitted more cases of `WriteBufferManager` triggering flushes when there are already immutable memtables awaiting flush, and (2) we didn't risk iterating the list of CFs in every write for many consecutive writes. To achieve both we could maintain a flag in `ColumnFamilyData` indicating whether the next switch would trigger write stall, and another flag in `DBImpl` indicating whether the previously-mentioned flag is true in every CF. We would only call `HandleWriteBufferFull()` when the `DBImpl` flag is false. Within `HandleWriteBufferFull()`, we would only trigger flush on a CF where the `ColumnFamilyData` flag is false.

If that's desirable or there's any other suggested change, I am happy to try it.

Reply:
ping @siying

Reply:
> (1) we permitted more cases of WriteBufferManager triggering flushes when there are already immutable memtables awaiting flush

This doesn't seem like a good idea to me anymore. Even if it won't lead to write stall, there's no need to trigger more flushes on the same CF.